### PR TITLE
feat: introduce the collector layer for tms

### LINKED ISSUE #10247: TMS - collector layer
The starting point for the collector layer is the `GH-gtm-collector` branch: https://github.com/SAP/spartacus/compare/feature/GH-gtm-collector

### PR DESCRIPTION
**Pull request by znikola**
closes #10247

### CODE REVIEW

File: projects/core/src/event/events.module.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { Inject, ModuleWithProviders, NgModule, Optional } from '@angular/core';
+import { EVENT_BUILDER } from './event.token';
+
+@NgModule({})
+export class EventsModule {
+  constructor(@Optional() @Inject(EVENT_BUILDER) _builders?: any[]) {
+    console.log('registered events: ', ..._builders);
```
Comment (dunqan) [inline projects/core/src/event/events.module.ts]: Do `console.log` proofs that it works as expected? ;-)

File: projects/core/src/event/tms.event.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import { createFrom } from '../util/create-from';
+import { EventService } from './event.service';
+
+export class TmsEvent {
+  static type = 'TmsEvent';
+  event: string;
+  payload: any;
+}
+
+export abstract class AbstractTmsEventCollector {
```
Comment (dunqan) [inline projects/core/src/event/tms.event.ts]: As it's abstract by definition, we don't have to say it in the name IMO.  so would go with just `TmsEventCollector`.

File: feature-libs/tms/core/gtm/gtm.service.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  protected subscription = new Subscription();
+
+  constructor(
+    protected eventsService: EventService,
+    protected windowRef: WindowRef
+  ) {}
+
+  collect(): void {
+    // prep data layer
+    this.window.dataLayer = this.window.dataLayer || [];
```
Comment (dunqan) [inline feature-libs/tms/core/gtm/gtm.service.ts]: we need some logic for SSR

File: feature-libs/tms/tsconfig.lib.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    "experimentalDecorators": true,
+    "importHelpers": true,
+    "types": [],
+    "lib": ["dom", "esnext"],
+    "paths": {
+      "@spartacus/core": ["dist/core"]
+    }
+  },
+  "angularCompilerOptions": {
+    "skipTemplateCodegen": true,
```
Comment (dunqan) [inline feature-libs/tms/tsconfig.lib.json:21]: What about adding `strictTemplates` in new libs, as it's the most strict AFAIK? 
(I know, it won't change here much, but as a convention).

File: feature-libs/tms/tsconfig.lib.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "outDir": "../../out-tsc/lib",
```
Comment (dunqan) [inline feature-libs/tms/tsconfig.lib.json:4]: What about the strict mode idea for new libs?

File: projects/core/src/auth/user-auth/services/auth-redirect-storage.service.ts [hunk 6]
```diff
@@ -11,17 +11,15 @@ import { BehaviorSubject, Observable } from 'rxjs';
...
+  private redirectUrl$ = new BehaviorSubject<string>(undefined);
 
   /**
    * Get redirect url after logging in.
    *
    * @returns observable with the redirect url as string
    */
   getRedirectUrl(): Observable<string> {
-    return this.redirectUrl$;
+    return this.redirectUrl$.asObservable();
```
Comment (dunqan) [inline projects/core/src/auth/user-auth/services/auth-redirect-storage.service.ts]: Why this change?

File: projects/core/src/cart/event/cart-event.collector.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    this.cartAddEntryEvent$,
+    this.cartAddEntrySuccessEvent$,
+    this.cartAddEntryFailEvent$,
+    this.cartRemoveEntrySuccessEvent$,
+    this.cartUpdateEntrySuccessEvent$,
+  ];
+
+  constructor(protected eventsService: EventService) {
+    super(eventsService);
+    this.registerEvent(...this.events$);
```
Comment (dunqan) [inline projects/core/src/cart/event/cart-event.collector.ts]: Will it work if moved to superclass?

File: projects/core/src/event/tms.event.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  protected abstract events$: Observable<TmsEvent>[];
```
Comment (dunqan) [inline projects/core/src/event/tms.event.ts]: Thinking about better name here... maybe `sources`? Also, it's an array, not an observable, so we should not use `$` here.

File: projects/core/src/cart/event/cart-event.collector.ts (hunk 7 above)
Comment (znikola) [reply projects/core/src/cart/event/cart-event.collector.ts]: Sadly, it won't.
I've tried it already, and this is what I got: 
Screen Shot 2021-01-11 at 09 55 29

File: projects/core/src/event/event.service.ts [hunk 9]
```diff
@@ -47,29 +48,42 @@ export class EventService {
...
-        console.warn(
-          `EventService: the event source`,
-          source$,
-          `has been already registered for the type`,
-          eventType
-        );
+  register<T>(eventType: AbstractType<T>, source$: Observable<T>): () => void {
+    const mergingSubjects: MergingSubject<T>[] = [];
+    let parentType = eventType;
+    while (
```
Comment (dunqan) [inline projects/core/src/event/event.service.ts]: I would expect the register method will be free from iterating through the prototype chain thing.

File: projects/core/src/event/event.service.ts [hunk 10]
```diff
@@ -47,29 +48,42 @@ export class EventService {
...
       }
-    } else {
-      eventMeta.mergingSubject.add(source$);
+
+      parentType = Object.getPrototypeOf(parentType);
     }
 
-    return () => eventMeta.mergingSubject.remove(source$);
+    return () =>
+      mergingSubjects.forEach((mergingSubject) =>
```
Comment (dunqan) [inline projects/core/src/event/event.service.ts]: This one also should not be needed IMO

File: projects/storefrontlib/src/events/events.module.ts [hunk 11]
```diff
@@ -1,9 +1,15 @@
 import { NgModule } from '@angular/core';
 import { CartPageEventModule } from './cart/cart-page-event.module';
+import { NavigationEventModule } from './navigation/navigation-event.module';
 import { PageEventModule } from './page/page-event.module';
 import { ProductPageEventModule } from './product/product-page-event.module';
 
 @NgModule({
-  imports: [CartPageEventModule, PageEventModule, ProductPageEventModule],
+  imports: [
+    NavigationEventModule,
```
Comment (dunqan) [inline projects/storefrontlib/src/events/events.module.ts]: As this module is deprecated already on develop, I'm wondering, if we should introduce changes to it.

File: projects/core/src/config/services/configuration.service.spec.ts [hunk 12]
```diff
@@ -26,6 +32,7 @@ describe('ConfigurationService', () => {
   beforeEach(() => {
     TestBed.configureTestingModule({
       providers: [
+        { provide: FeatureConfigService, useClass: MockFeatureConfigService },
```
Comment (dunqan) [inline projects/core/src/config/services/configuration.service.spec.ts]: Why we need this?

File: projects/core/src/event/event.service.ts [hunk 13]
```diff
@@ -42,12 +48,14 @@ export class EventService {
...
    *
+   * @since 3.1 - registers the given `source$` for the parent classes of the given `eventType`.
+   *
    * @param eventType the event type
    * @param source$ an observable that represents the source
    *
    * @returns a teardown function which unregisters the given event source
    */
-  register<T>(eventType: Type<T>, source$: Observable<T>): () => void {
+  register<T>(eventType: AbstractType<T>, source$: Observable<T>): () => void {
```
Comment (dunqan) [inline projects/core/src/event/event.service.ts:61]: This one should be changed to `Type` in the long run, so if not now because of some breaking changes, then we should leave some note/TODO here.

File: feature-libs/tms/public_api.ts [hunk 14]
```diff
@@ -0,0 +1,5 @@
+/*
+ * Public API Surface of tms
+ */
+
+export * from './core/index';
```
Comment (dunqan) [inline feature-libs/tms/public_api.ts]: I would expect the structure to look like this:
```
@spartacus/tms/core
@spartacus/tms/gtm
@spartacus/tms/adobelaunch
```
Any opinions?

File: projects/core/src/config/services/configuration.service.spec.ts (hunk 12 above)
Comment (znikola) [reply projects/core/src/config/services/configuration.service.spec.ts]: The test was breaking without it.

File: projects/core/src/event/event.service.ts [hunk 15]
```diff
@@ -104,7 +112,7 @@ export class EventService {
   /**
    * Returns the event meta object for the given event type
    */
-  private getEventMeta<T>(eventType: Type<T>): EventMeta<T> {
+  private getEventMeta<T>(eventType: AbstractType<T>): EventMeta<T> {
     if (isDevMode()) {
       this.validateEventType(eventType);
```
Comment (dunqan) [inline projects/core/src/event/event.service.ts]: Wondering, if we should not move this into create or put in the `if` block below, to not have the warning with every get, but only once (at first get) ?

File: projects/core/src/event/event.service.ts (hunk 13 above)
Comment (znikola) [reply projects/core/src/event/event.service.ts:61]: Added TODO

File: projects/core/src/event/event.service.ts [hunk 16]
```diff
@@ -115,27 +123,63 @@ export class EventService {
...
-  private validateEventType<T>(eventType: Type<T>): void {
+  private validateEventType<T>(eventType: AbstractType<T>): void {
     if (!eventType?.constructor) {
       throw new Error(
         `EventService:  ${eventType} is not a valid event type. Please provide a class reference.`
       );
     }
+
+    if (this.featureConfigService?.isLevel('3.1')) {
+      this.validateCxEvent(eventType);
```
Comment (dunqan) [inline projects/core/src/event/event.service.ts:162]: not sure if feature flag is really needed for this check

File: projects/storefrontlib/src/events/events.module.ts (hunk 11 above)
Comment (znikola) [reply projects/storefrontlib/src/events/events.module.ts]: Added to `projects/storefrontlib/src/recipes/storefront-foundation.module.ts` instead

File: projects/core/src/event/event.service.ts [hunk 17]
```diff
@@ -115,27 +123,63 @@ export class EventService {
...
+  private createEventMeta<T>(eventType: AbstractType<T>): void {
+    const eventMeta: EventMeta<T> = {
       inputSubject$: null, // will be created lazily by the `dispatch` method
-      mergingSubject: new MergingSubject(),
-    });
+      mergingSubject: new MergingSubject<T>(),
+    };
+    this.eventsMeta.set(eventType, eventMeta);
+
+    if (this.featureConfigService?.isLevel('3.1')) {
```
Comment (dunqan) [inline projects/core/src/event/event.service.ts:136]: any TODOs to remove this in 4.0?

File: projects/storefrontlib/src/events/page/page-event.builder.ts [hunk 18]
```diff
@@ -1,58 +1,55 @@
...
+    protected eventService: EventService,
+    /** @deprecated @since 3.1 - this will be remove in 4.0 */ protected featureConfigService?: FeatureConfigService
   ) {
     this.register();
   }
 
   protected register(): void {
-    this.eventService.register(PageEvent, this.buildPageEvent());
+    if (this.featureConfigService?.isLevel('!3.1')) {
+      this.eventService.register(PageEvent, this.buildPageEvent());
```
Comment (dunqan) [inline projects/storefrontlib/src/events/page/page-event.builder.ts:32]: Wouldn't it be more clear to have separate HomePageEventBuilder, so this big one would be deprecated and removed in 4.0?

File: feature-libs/tms/public_api.ts (hunk 14 above)
Comment (znikola) [reply feature-libs/tms/public_api.ts]: Foundation done.
We can discuss it further (ll, core, root, etc.)

File: feature-libs/tms/adobe-launch/services/adobe-launch.service.ts [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    protected eventsService: EventService,
+    protected windowRef: WindowRef,
+    protected tmsConfig: TmsConfig
+  ) {}
+
+  /**
+   * This method is called only once to start collecting and dispatching events to Adobe Launch
+   */
+  collect(): void {
+    this.prepareDataLayer();
```
Comment (dunqan) [inline feature-libs/tms/adobe-launch/services/adobe-launch.service.ts:36]: Should we run this logic on a server side?

File: feature-libs/tms/gtm/services/gtm.service.ts [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /**
+   * Called each time an event fires.
+   *
+   * @param event Spartacus event to dispatch
+   */
+  protected pushToDataLayer<T extends CxEvent>(event: T): void {
+    if (isDevMode()) {
+      console.log(`🎤 GTM received event: ${JSON.stringify(event)}`);
```
Comment (dunqan) [inline feature-libs/tms/gtm/services/gtm.service.ts:72]: Is this meant to be in the final code?

File: feature-libs/tms/gtm/services/gtm.service.ts [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    protected eventsService: EventService,
+    protected windowRef: WindowRef,
+    protected tmsConfig: TmsConfig
+  ) {}
+
+  /**
+   * This method is called only once to start collecting and dispatching events to GTM
+   */
+  collect(): void {
+    this.prepareDataLayer();
```
Comment (dunqan) [inline feature-libs/tms/gtm/services/gtm.service.ts:36]: I guess we should not run it on the sever side

File: projects/core/src/event/event.service.ts (hunk 17 above)
Comment (znikola) [reply projects/core/src/event/event.service.ts:136]: Added

File: projects/core/src/event/event.service.ts (hunk 16 above)
Comment (znikola) [reply projects/core/src/event/event.service.ts:162]: Not sure what do you mean?
The validation method should be called if we use proto chain to register events.

File: projects/core/src/event/event.service.ts (hunk 15 above)
Comment (znikola) [reply projects/core/src/event/event.service.ts]: Nice, I moved it a few lines below.

File: feature-libs/tms/gtm/services/gtm.service.ts (hunk 21 above)
Comment (znikola) [reply feature-libs/tms/gtm/services/gtm.service.ts:36]: It's not running, as it checks for the window inside of it.

File: feature-libs/tms/adobe-launch/services/adobe-launch.service.ts (hunk 19 above)
Comment (znikola) [reply feature-libs/tms/adobe-launch/services/adobe-launch.service.ts:36]: see the comment below: https://github.com/SAP/spartacus/pull/10629#discussion_r565199515

File: feature-libs/tms/gtm/services/gtm.service.ts (hunk 20 above)
Comment (znikola) [reply feature-libs/tms/gtm/services/gtm.service.ts:72]: It can be :)
It will be overriden by customers anyways, and it's useful to distinguish the messages in the console.

File: projects/storefrontlib/src/events/page/page-event.builder.ts (hunk 18 above)
Comment (znikola) [reply projects/storefrontlib/src/events/page/page-event.builder.ts:32]: Done

File: projects/storefrontlib/src/events/events.module.ts (hunk 11 above)
Comment (dunqan) [reply projects/storefrontlib/src/events/events.module.ts]: Its also deprecated ;-)

File: projects/core/src/event/event.service.ts (hunk 16 above)
Comment (dunqan) [reply projects/core/src/event/event.service.ts:162]: I understand it's for checking if event extends cxEvent?

File: feature-libs/tms/gtm/services/gtm.service.ts (hunk 21 above)
Comment (dunqan) [reply feature-libs/tms/gtm/services/gtm.service.ts:36]: Checking for the window is not a good enough check for SSR (as we realized in the past)

File: feature-libs/tms/gtm/services/gtm.service.ts (hunk 20 above)
Comment (dunqan) [reply feature-libs/tms/gtm/services/gtm.service.ts:72]: let's resolve this one here, but have a final agreement with some other input (maybe on a blamed catch-up)

File: feature-libs/tms/gtm/services/gtm.service.ts (hunk 21 above)
Comment (znikola) [reply feature-libs/tms/gtm/services/gtm.service.ts:36]: You're right, done.

File: feature-libs/tms/gtm/services/gtm.service.ts (hunk 20 above)
Comment (znikola) [reply feature-libs/tms/gtm/services/gtm.service.ts:72]: Added the debug config

File: projects/core/src/event/event.service.ts (hunk 16 above)
Comment (znikola) [reply projects/core/src/event/event.service.ts:162]: Yes

File: projects/storefrontlib/src/events/events.module.ts (hunk 11 above)
Comment (znikola) [reply projects/storefrontlib/src/events/events.module.ts]: Yes, I realized that later. Move to _spartacus feature module_

**Review by dunqan — APPROVED** [review]
